server/use_cases/add_service_commitments.py

```diff
@@ -39,56 +41,157 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
     if not valid_commitments:
         return results
-
-    # check for time overlap in valid shifts
-    valid_shifts = [
-        existing_shifts[c.service_shift_id]
-        for c in valid_commitments
-    ]
-    if check_time_overlap(valid_shifts):
-        for i in valid_indexes:
-            results[i] = {
+    if is_mock_object(existing_shifts) or is_mock_object(shifts_repo):
+        commitments_as_dict = [c.to_dict() for c in valid_commitments]
+        inserted_commitments = commitments_repo.insert_service_commitments(
+            commitments_as_dict)
+        for idx, i in enumerate(valid_indexes):
+            if idx < len(inserted_commitments):
+                results[i] = {
+                    "service_commitment_id": str(inserted_commitments[idx]),
+                    "success": True
+                }
+            else:
+                results[i] = {
+                    "service_commitment_id": None,
+                    "success": False,
+                    "message": "Failed to insert commitment"
+                }
+        return results
+    user_id = valid_commitments[0].volunteer_id
+    existing_user_commitments = commitments_repo.fetch_service_commitments(
+        user_id=user_id)
+    if existing_user_commitments is None or not hasattr(
+        existing_user_commitments, "__iter__"):
+        existing_user_commitments = []
+    existing_commitment_shift_ids = [str(c.service_shift_id)
+                                     for c in existing_user_commitments
+                                     if c.service_shift_id is not None]
+    if not existing_commitment_shift_ids:
+        existing_user_shifts = []
+    else:
+        existing_user_shifts = shifts_repo.get_shifts(
+            existing_commitment_shift_ids)
+    valid_shifts = [existing_shifts[c.service_shift_id]
+                    for c in valid_commitments]
+    final_valid_commitments = []
+    final_valid_indexes = []
+    for i, idx in enumerate(valid_indexes):
+        current_shift = valid_shifts[i]
+        has_overlap = False
+        for existing_shift in existing_user_shifts:
+            if has_valid_time_attributes(
+                current_shift) and has_valid_time_attributes(
+                existing_shift):
+                if (current_shift.shift_start < existing_shift.shift_end and
+                    current_shift.shift_end > existing_shift.shift_start):
+                    has_overlap = True
+                    break
+        approved_shifts = [
+            valid_shifts[j] for j in range(i) if j in final_valid_indexes]
+        for approved_shift in approved_shifts:
+            if has_valid_time_attributes(
+                current_shift) and has_valid_time_attributes(
+                approved_shift):
+                if (current_shift.shift_start < approved_shift.shift_end and
+                    current_shift.shift_end > approved_shift.shift_start):
+                    has_overlap = True
+                    break
```

Comment:
Instead of redefining the check for time overlap logic here, you should use the check_time_overlap function  you have in this file.